File: dynamic-programming/leetcode53.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
"""
    @Header leetcode53
    @Author fosaken
    @Date   2020/12/15
    @Abstract:
        最大子序和
        Address: https://leetcode-cn.com/problems/maximum-subarray/
"""

from typing import List


def maxSubArray(nums: List[int]) -> int:
    l = len(nums)
    s = 0
    max_s = float('-inf')
    for i in range(l):
        s += nums[i]
        max_s = max(max_s, s)
        if s < 0:
            s = 0
    return max_s

File: dynamic-programming/leetcode53.cpp
#include <iostream>
#include <vector>

using namespace std;

int maxSubArray(vector<int>& nums){
    vector<int> dp(nums.begin(), nums.end());
    for (int i = 1; i < dp.size(); i++){
        if(dp[i-1] > 0){
            dp[i] += dp[i - 1];
        }
    }
    int res = INT_MIN;
    for (int i = 0; i < dp.size(); i++){
        res = max(res, dp[i]);
    }
    return res;
}
int main(){
    vector<int> nums = {-2,1,-3,4,-1,2,1,-5,4};
    cout << maxSubArray(nums);
    system("pause");
    return 0;
}